[REVIEW]gQuant UI, first version

yidong72 opened the pull request:
<!--

Thank you for contributing to gQuant :)

Here are some guidelines to help the review process go smoothly.

1. Please write a description in this text box of the changes that are being
   made.

2. Please ensure that you have written units tests for the changes made/features
   added.

3. There are CI checks in place to enforce that committed code follows our style
   and syntax standards. Please see our contribution guide in `CONTRIBUTING.MD`
   in the project root for more information about the checks we perform and how
   you can run them locally.

4. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

5. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]`.

6. If your pull request is ready to be reviewed without requiring additional
   work on top of it, then remove the `[WIP]` label (if present) and replace
   it with `[REVIEW]`. If assistance is required to complete the functionality,
   for example when the C/C++ code of a feature is complete but Python bindings
   are still required, then add the label `[HELP-REQ]` so that others can triage
   and assist. The additional changes then can be implemented on top of the
   same PR. If the assistance is done by members of the rapidsAI team, then no
   additional actions are required by the creator of the original PR for this,
   otherwise the original author of the PR needs to give permission to the
   person(s) assisting to commit to their personal fork of the project. If that
   doesn't happen then a new PR based on the code of the original PR can be
   opened by the person assisting, which then will be the PR that will be
   merged.

7. Once all work has been done and review has taken place please do not add
   features or make changes out of the scope of those requested by the reviewer
   (doing this just add delays as already reviewed code ends up having to be
   re-reviewed/it is hard to tell what is new etc!). Further, please do not
   rebase your branch on master/force push/rewrite history, doing any of these
   causes the context of any comments made by reviewers to be lost. If
   conflicts occur against master they should be resolved by merging master
   into the branch used for making the pull request.

Many thanks in advance for your cooperation!

-->
In this PR, it added

- Node configuration JSON schema, so the conf can be validated. It also helps the client to edit the conf file.
- flask backend server to feed node information to the client 
- React client web UI. It has the feature of pan, zoom the graph, add and edit the nodes, make connections between the ports, download the workflow, automatically layout the graph.

yidong72 commented:
I did the rebase as you suggested.

yidong72 commented:
Checked in the last feature for this PR. When making new connections, it can dynamically show who are the valid ports for this connection.

avolkov1 commented on client/ui/.vscode/launch.json:
```diff
@@ -0,0 +1,15 @@
+{
```
The folder structure should be organized to clearly separate the client and server. Currently you have the backend and frontend under one directory structure "client", but client implies primarily frontend.

Maybe have something like:
```
<gQuant repo>
| - gquantui - rename current "client" to gquantui or something similar
|   | - client - rename current "ui" to just client or "gquantui-client"
|   |   | - .vscode - corresponds to current "<gQuant repo>/client/ui/.vscode"
|   |   \ - all the other files under "ui" currently
|   |
|   | - server - or similar name such as "gquantui-server"
|   |   | - .vcode - corresponds to current "<gQuant repo>/.vscode"
|   |   \ - server.py and other files for server

```

avolkov1 commented on .vscode/launch.json:
```diff
@@ -0,0 +1,29 @@
+{
```
The folder structure should be organized to clearly separate the client and server. Currently you have the backend and frontend under one directory structure "client", but client implies primarily frontend.

Maybe have something like:
```
<gQuant repo>
| - gquantui - rename current "client" to gquantui or something similar
|   | - client - rename current "ui" to just client or "gquantui-client"
|   |   | - .vscode - corresponds to current "<gQuant repo>/client/ui/.vscode"
|   |   \ - all the other files under "ui" currently
|   |
|   | - server - or similar name such as "gquantui-server"
|   |   | - .vcode - corresponds to current "<gQuant repo>/.vscode"
|   |   \ - server.py and other files for server

```

This top-level ".vscode" launcher configuration should be in a server sub-directory.

avolkov1 commented on client/static/js/app.js:
```diff
@@ -0,0 +1,360 @@
+window.onload = (event) => {
+d3.json('/graph').then((d)=>{
```
Should this be:
```
d3.json(process.env.REACT_APP_GRAPH_URL).then((d)=>{
```

Using `process.env.REACT_APP_GRAPH_URL` instead of direct url `/graph`

GPUtester commented:
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

yidong72 commented on client/README.md:
````diff
@@ -0,0 +1,36 @@
+
+```bash
+cd ui
+npm install
````
done

yidong72 commented on client/README.md:
````diff
@@ -0,0 +1,36 @@
+the server API and the static built page are hosted by Flask
+
+```bash
+python server.py
````
The  README is fully updated

yidong72 commented on client/server.py:
```diff
@@ -0,0 +1,39 @@
+
+@app.route('/graph')
+def get_graph():
+    filename = str(pathlib.Path(__file__).parent / 'custom_wflow.yaml')
```
the file selection box is added

yidong72 commented on client/server.py:
```diff
@@ -0,0 +1,39 @@
+
+@app.route('/add')
+def add_node():
+    result = add_nodes('custom_port_nodes')
```
It automatically load all the modules in a sub-directory 'modules'

yidong72 commented on client/server.py:
```diff
@@ -0,0 +1,39 @@
+import sys; sys.path.insert(0, '..')
+from flask import Flask, jsonify, request
+from gquant.dataframe_flow import TaskGraph
+from get_graph import get_nodes, add_nodes, get_nodes_from_file
```
changed the name to 'server_utils'

yidong72 commented on client/get_graph.py:
```diff
@@ -0,0 +1,96 @@
+import uuid
+
+
+def get_node_obj(node):
```
added

yidong72 commented on client/get_graph.py:
```diff
@@ -0,0 +1,96 @@
+    return {'nodes': nodes, 'edges': edges}
+
+
+def add_nodes(module):
```
added

yidong72 commented on client/get_graph.py:
```diff
@@ -0,0 +1,96 @@
+    return get_nodes(task_graph)
+
+
+def get_nodes(task_graph):
```
added

yidong72 commented on client/get_graph.py:
```diff
@@ -0,0 +1,96 @@
+    return out_node
+
+
+def get_nodes_from_file(file):
```
added

yidong72 commented:
I addressed all your comments. I added a workspace file so that both server and client project can be simultaneously opened up in the VSCode inside the container. I put the detailed instructions in the README file. I added the file manager to load the yaml files.

yidong72 commented:
I did a few changes:

1. Move the file widget to the left 
2. Feed the port type information to the client. The validation function also checks whether the port type matches or not.
3. allow user to edit the filename for download and save
4. Improves the performance. no more lagging experiences

avolkov1 commented:
Please hold off on merging other PRs into this PR until the other PRs are closed. This confuses the commits history (making it hard to review), will cause rebase conflicts which can cause bugs fixes in other PRs not getting merged properly. This branch will need to be rebased to clean up the commits and make sure to fix bugs during merge conflicts.

avolkov1 commented:
Let's clean up this branch. I would make a copy backup of the local repo directory while in `branch-client` just in case something goes missing.

Use the following commands:
```
$ git checkout branch-client
# ghub-gquant is my remote name for: git@github.com:rapidsai/gQuant.git
$ git rebase -Xours ghub-gquant/develop
$ git rebase  -i 5d53970
```
During `rebase -i 5d53970` you'll see a list of commits. Keep only the following ones:
```
pick 0884fc8 make the code 0.13 compatible
pick 73f394c :# This is a combination of 2 commits.
pick 1f7db16 fixed the flake error
pick 5d8c9b6 feed the type to the client
pick e34598f now check the types of the port
pick 20035d2 fixed the missing line issue
pick 3972ebf use firefox to debug
pick 50f2a2e save the workflow in disk
pick b66ce5d try new workspace
```

The SHAs might differ. I believe the above ones are the only relevant ones for this branch. Delete the other lines. Doing above I did get any conflicts and commits from develop were properly applied.

Then do a push force.
```
$ git push --force
```

yidong72 commented:
> 
> 
> Let's clean up this branch. I would make a copy backup of the local repo directory while in `branch-client` just in case something goes missing.
> 
> Use the following commands:
> 
> ```
> $ git checkout branch-client
> # ghub-gquant is my remote name for: git@github.com:rapidsai/gQuant.git
> $ git rebase -Xours ghub-gquant/develop
> $ git rebase  -i 5d53970
> ```
> 
> During `rebase -i 5d53970` you'll see a list of commits. Keep only the following ones:
> 
> ```
> pick 0884fc8 make the code 0.13 compatible
> pick 73f394c :# This is a combination of 2 commits.
> pick 1f7db16 fixed the flake error
> pick 5d8c9b6 feed the type to the client
> pick e34598f now check the types of the port
> pick 20035d2 fixed the missing line issue
> pick 3972ebf use firefox to debug
> pick 50f2a2e save the workflow in disk
> pick b66ce5d try new workspace
> ```
> 
> The SHAs might differ. I believe the above ones are the only relevant ones for this branch. Delete the other lines. Doing above I did not get any conflicts and commits from develop were properly applied.
> 
> Then do a push force.
> 
> ```
> $ git push --force
> ```

Thanks. I follwed the instruction and fixed the commits in this branch. It works fine.

yidong72 commented:
I finished the work of migrating the UI to Jupyterlab ecosystem. The user can select a set of output ports from the UI and run it in a notebook cell. Check the `test.ipynb` notebook for example.

avolkov1 commented on gquant/dataframe_flow/taskGraph.py:
```diff
@@ -388,20 +523,21 @@ def process(self, inputs):
             i.flow()
 
         results_dfs_dict = outputs_collector_node.input_df
+        Result = namedtuple('Result', " ".join(outputs).replace('.', '_'))
```
It's a bit awkward to rename the outputs (inconsistent with notation we're using `<node_id>.<out_port>`). I was testing various ways to achieve the tuple unpacking API, but be able to have a dict like container. Maybe just overriding a dict (or OrderedDict):
```
class Results(OrderedDict): 
    def __iter__(self): 
        return iter(self.values())

mdict = Results([('x.y', 5), ['u.v', 6]])

print(mdict)
Results([('x.y', 5), ('u.v', 6)])

In [61]: x, y = mdict                                                                                                                                                                                              

In [64]: x, y                                                                                                                                                                                                      
Out[64]: (5, 6)

n [65]: def tfn(tvar): 
    ...:     return tvar 
    ...:                                                                                                                                                                                                           

In [66]: (x,y) = tfn(mdict)                                                                                                                                                                                          

In [67]: x, y                                                                                                                                                                                                      
Out[67]: (5, 6)

```

Above behaves like a dict, but unpacks the values instead of keys. That will make it compatible with tuple API in existing notebooks. Otherwise it can be used like a dictionary.

yidong72 replied:
I created a Result object to have all the desirable behaviors we want
```python
class Results(object):

    def __init__(self, values):
        self.values = tuple([i[1] for i in values])
        self.__dict = OrderedDict(values)

    def __iter__(self):
        return iter(self.values)

    def __getitem__(self, key):
        if isinstance(key, int):
            return self.values[key]
        else:
            return self.__dict[key]

    def __len__(self):
        return len(self.values)

    def __repr__(self):
        return "Results"+self.__dict.__repr__()[11:]

    def __str__(self):
        return "Results"+self.__dict.__str__()[11:]

    def __contains__(self, key):
        return True if key in self.__dict else False
```
you can do
```python
x, y = mdict 
mydict[0]
mydict[1]
mydict['x.y']
mydict['u.v']
```

avolkov1 replied:
Looks good.

avolkov1 commented on gquant/dataframe_flow/task.py:
```diff
@@ -67,6 +68,7 @@ def get_node_obj(self, replace=None, profile=False, tgraph_mixin=False):
                 spec = importlib.util.spec_from_file_location(
                     modulename, modulepath)
                 mod = importlib.util.module_from_spec(spec)
+                sys.modules[spec.name] = mod
```
I get an error when running the `test_dask.ipynb` notebook.
```
  File "<my conda python>/lib/python3.7/site-packages/distributed/protocol/pickle.py", line 75, in loads
    return pickle.loads(x)
ModuleNotFoundError: No module named 'custom_port_nodes'
```

To fix this I've modified this part for dask logic to add the path on the workers.
```
            if modulepath is not None:
                spec = importlib.util.spec_from_file_location(node_id,
                                                              modulepath)
                pp = Path(modulepath)
                modulename = pp.stem
                spec = importlib.util.spec_from_file_location(
                    modulename, modulepath)
                mod = importlib.util.module_from_spec(spec)
                sys.modules[spec.name] = mod

                module_dir = str(pp.parent.absolute())

                def append_path(path):
                    if path not in sys.path:
                        sys.path.append(path)

                append_path(module_dir)

                try:
                    import dask.distributed
                    client = dask.distributed.client.default_client()
                    client.run(append_path, module_dir)
                except (ValueError, ImportError):
                    pass

                spec.loader.exec_module(mod)
                NodeClass = getattr(mod, node_type)
            else:
                . . .
```

Do you have this problem of dask not being able to find `custom_port_nodes` module? If I put `custom_port_nodes.py` in the same directory as the notebook then the notebook works fine.

yidong72 replied:
Good catch. I took your code and fixed it.

avolkov1 commented on gquant/dataframe_flow/taskGraph.py:
```diff
@@ -414,5 +630,12 @@ def draw(self, show=None, fmt='png', show_ports=False):
 
             plt = Image(pdot_out)
             display(plt)
+        elif show in ('lab',):
+            from gquantlab.gquantmodel import GQuantWidget
+            widget = GQuantWidget()
+            widget.value = self.export_task_speclist()
+            widget.set_taskgraph(self)
+            self.__widget = widget
+            return widget
```
I think this logic should change a bit. The `get_widget` should be widget creator or maybe create a separate method to create a widget. The draw for lab should display the widget.
```
    def get_widget(self):
        if self.__widget is None:
            from gquantlab.gquantmodel import GQuantWidget
            widget = GQuantWidget()
            widget.value = self.export_task_speclist()
            widget.set_taskgraph(self)
            self.__widget = widget

        return self.__widget

```

In the draw method:
```
        elif show in ('lab',):
            from IPython.display import display
            widget = self.get_widget()
            display(widget)
            return widget
```

yidong72 replied:
Agree on the `get_widget` being a widget creator.
I don't think we need to call that "display" function. It breaks the behavior of ipywidgets. 
if you do this
```python
import ipywidgets
a = ipywidgets.HTML('<h1>hello</h1>')
```
the output cell is empty as assignment operation has not returned value.
It only shows the content in the output cell when you call
```
ipywidgets.HTML('<h1>hello</h1>')
```

avolkov1 replied:
I see. Maybe the draw method shouldn't return the widget for `show='lab'` option.

That would be more consistent with `show='ipynb'` option.

The lab widget can be retrieved via `get_widget` (widget is too generic, I think it should be renamed to be more descriptive, maybe `get_quantlab`).

The pdot part is also irrelevant for the lab option.
```
    def draw(self, show='lab', fmt='png', show_ports=False):
        if show in ('ipynb',):
            pdot = self.to_pydot(show_ports)
            pdot_out = pdot.create(format=fmt)
            from IPython.display import display
            if fmt in ('svg',):
                from IPython.display import SVG as Image  # @UnusedImport
            else:
                from IPython.display import Image  # @Reimport

            plt = Image(pdot_out)
            display(plt)
        elif show in ('lab',):
            from IPython.display import display
            widget = self.get_widget()
            display(widget)
        
```

That would be one less cell in Jupyter to have to display a gquantlab widget.

I'm not sure if above breaks the sync states between widget references in other cells. I did this out of curiosity and there was some unexpected sync behavior with outputs collector node.

avolkov1 replied:
These changes make the `draw` behave consistently so that fixes this issue. I'll mark resolved.

yidong72 commented:
Addressed your two comments

yidong72 commented:
I switched to the `configparser` library, added the optional 'module' field to the node so that it can load the exact class from the specified module name. If no 'module' is set, it will scan through all the modules specified in the `gquantrc` file. I also default the `GQUANT_PATH` to be `os.getcwd()+/gquantrc` and `MODULEPATH` to be `os.getcwd()+/modules/`

yidong72 commented:
all the gQuant nodes are migrated

yidong72 commented:
Issue #95 and #97 are all addressed in this PR. It is ready to review and merge

avolkov1 commented on docker/build.sh:
```diff
@@ -1,111 +1,165 @@
-    && chmod 777 /.local /.jupyter /.config /.cupy
-RUN source activate rapids \ 
-    && jupyter lab build && jupyter lab clean
+FROM nvidia/cuda:$OS_STR
```
I have this modified slightly. 

Need to decouple `OS_STR` from CUDA runtime. The `OS_STR` is defined as:
```
case $OPERATING_SYSTEM in
    1)
        echo "Ubuntu 16.04 selected."
        OS_STR="ubuntu16.04"
        ;;
    2)
        echo "Ubuntu 18.04 selected."
        OS_STR="ubuntu18.04"
        ;;
    *)
        echo "Ubuntu 20.04 selected."
        OS_STR="ubuntu20.04"
        ;;
esac
```



And then the container name/tag should include gquant version. I have it defined as:
```
gquant_ver=$(grep version gQuant/setup.py | sed "s/^.*version='\([^;]*\)'.*/\1/")
CONTAINER="nvidia/cuda:${CUDA_STR}-runtime-${OS_STR}"
D_CONT=${D_CONT:="gquant/gquant:${gquant_ver}-${CUDA_STR}_${OS_STR}_${RAPIDS_VERSION}_${MODE_STR}"}

cat > $D_FILE <<EOF
FROM $CONTAINER
```

avolkov1 replied:
https://github.com/yidong72/gQuant/blob/757d027d0a87cacb46b51a298716fd27ee0c77d8/docker/build.sh#L28

Typo. Please fix to:
```
        OS_STR="ubuntu20.04"
```

https://github.com/yidong72/gQuant/blob/757d027d0a87cacb46b51a298716fd27ee0c77d8/docker/build.sh#L53-L54
Typo. Please fix echo to:
```
        echo "CUDA 11.0 is selected"
```

https://github.com/yidong72/gQuant/blob/757d027d0a87cacb46b51a298716fd27ee0c77d8/docker/build.sh#L90-L92
These three lines should come after the copying of gQuant to local directory, because the `grep version gQuant/setup.py` looks inside that gQuant subdirectory. Change to:
```
mkdir -p gQuant
cp -r ../gquant ./gQuant
cp -r ../task_example ./gQuant
cp -r ../modules ./gQuant
cp -r ../taskgraphs ./gQuant
cp ../setup.cfg ./gQuant
cp ../setup.py ./gQuant
cp ../LICENSE ./gQuant
cp ../download_data.sh ./gQuant
cp ../gquantrc ./gQuant
rsync -av --progress ../notebooks ./gQuant --exclude data --exclude .cache --exclude many-small --exclude storage --exclude dask-worker-space --exclude __pycache__
rsync -av --progress ../gquantlab ./gQuant --exclude node_modules 

gquant_ver=$(grep version gQuant/setup.py | sed "s/^.*version='\([^;]*\)'.*/\1/")
CONTAINER="nvidia/cuda:${CUDA_STR}-runtime-${OS_STR}"
D_CONT=${D_CONT:="gquant/gquant:${gquant_ver}-${CUDA_STR}_${OS_STR}_${RAPIDS_VERSION}_${MODE_STR}"}

```

yidong72 replied:
good catch. I fixed them

avolkov1 commented on gquant/dataframe_flow/_node_flow.py:
```diff
@@ -4,14 +4,17 @@
-OUTPUT_ID = 'f291b900-bd19-11e9-aca3-a81e84f29b0f_uni_output'
+# OUTPUT_ID = 'f291b900-bd19-11e9-aca3-a81e84f29b0f_uni_output'
+OUTPUT_ID = 'collector_id_fd9567b6'
+OUTPUT_TYPE = 'Output Collector'
```
Best practice is to avoid having a space in a string identifier. Could you change it to `Output_Collector` using an underscore?

I found this identifier in typescript code, taskgraphs and some of the notebooks.

avolkov1 replied:
https://github.com/yidong72/gQuant/blob/757d027d0a87cacb46b51a298716fd27ee0c77d8/docker/build.sh#L28

Typo. Please fix to:
```
        OS_STR="ubuntu20.04"
```

https://github.com/yidong72/gQuant/blob/757d027d0a87cacb46b51a298716fd27ee0c77d8/docker/build.sh#L53-L54
Typo. Please fix echo to:
```
        echo "CUDA 11.0 is selected"
```

https://github.com/yidong72/gQuant/blob/757d027d0a87cacb46b51a298716fd27ee0c77d8/docker/build.sh#L90-L92
These three lines should come after the copying of gQuant to local directory, because the `grep version gQuant/setup.py` looks inside that gQuant subdirectory. Change to:
```
mkdir -p gQuant
cp -r ../gquant ./gQuant
cp -r ../task_example ./gQuant
cp -r ../modules ./gQuant
cp -r ../taskgraphs ./gQuant
cp ../setup.cfg ./gQuant
cp ../setup.py ./gQuant
cp ../LICENSE ./gQuant
cp ../download_data.sh ./gQuant
cp ../gquantrc ./gQuant
rsync -av --progress ../notebooks ./gQuant --exclude data --exclude .cache --exclude many-small --exclude storage --exclude dask-worker-space --exclude __pycache__
rsync -av --progress ../gquantlab ./gQuant --exclude node_modules 

gquant_ver=$(grep version gQuant/setup.py | sed "s/^.*version='\([^;]*\)'.*/\1/")
CONTAINER="nvidia/cuda:${CUDA_STR}-runtime-${OS_STR}"
D_CONT=${D_CONT:="gquant/gquant:${gquant_ver}-${CUDA_STR}_${OS_STR}_${RAPIDS_VERSION}_${MODE_STR}"}

```

yidong72 replied:
fixed

yidong72 commented on modules/custom_port_nodes.py:
```diff
@@ -0,0 +1,480 @@
+            # df2_col = df2_col.compute()
+            pass
+
+        max_difference = (df1_col - df2_col).abs().max()
```
Done with those changes

avolkov1 reviewed (commented):
All these changes resolved most of the comments.

There were a couple comments that I think you missed.

1. I wasn't sure if GQUANTROOT would ever be modified.
https://github.com/rapidsai/gQuant/pull/89#discussion_r480271163

2. build_dev.sh reference
https://github.com/rapidsai/gQuant/pull/89#discussion_r480265702

3. Docstrings and URL updates
https://github.com/rapidsai/gQuant/pull/89#discussion_r480273550
https://github.com/rapidsai/gQuant/pull/89#discussion_r480275422
https://github.com/rapidsai/gQuant/pull/89#discussion_r480275861

yidong72 commented on gquantlab/gquantlab/handlers.py:
```diff
@@ -0,0 +1,55 @@
+def setup_handlers(web_app):
+    host_pattern = ".*$"
+    base_url = web_app.settings["base_url"]
+    os.environ['GQUANTROOT'] = os.getcwd()
```
You are right. `os.getcwd()` is just the default behavior

yidong72 replied:
I make a close look at the GQUANTROOT variable. It can only be os.getcwd() as it is used to send the gquantlab working directory information to some utility functions to help to find the path of the files in the disk

yidong72 commented on gquantlab/gquantlab/server_utils.py:
```diff
@@ -0,0 +1,224 @@
+    """
+    It will load all the nodes for the UI client so user can add new node
+    to the graph. The nodes are from two sources:
+    default gQuant nodes and customized node modules. Currently gQuant default
+    nodes are still using old API. So all the nodes are coming from customized
+    node modules in the 'modules' directory.
```
done

yidong72 commented on gquantlab/package.json:
```diff
@@ -0,0 +1,94 @@
+    "jupyterlab",
+    "jupyterlab-extension"
+  ],
+  "homepage": "https://github.com/rapidsai/gQuant/gquantlab",
+  "bugs": {
+    "url": "https://github.com/rapidsai/gQuant/gquantlab/issues"
```
done

yidong72 commented on gquantlab/package.json:
```diff
@@ -0,0 +1,94 @@
+  "style": "style/index.css",
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/rapidsai/gQuant/gquantlab.git"
```
done

yidong72 commented on gquantlab/setup.py:
```diff
@@ -0,0 +1,87 @@
+setup_args = dict(
+    name=name,
+    version=version,
+    url="https://github.com/rapidsai/gQuant/gquantlab",
```
done

yidong72 commented:
sorry. For some reason I missed a few comments in the last review. I just went through them all

avolkov1 reviewed (approved):
Looks good. I'll squash merge this.
